hedera-node/src/main/java/com/hedera/services/store/schedule/HederaScheduleStore.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * @author Daniel Ivanov
+ */
+public class HederaScheduleStore extends HederaStore implements ScheduleStore {
+	static final ScheduleID NO_PENDING_ID = ScheduleID.getDefaultInstance();
+
+	private final Supplier<FCMap<MerkleEntityId, MerkleSchedule>> schedules;
+	Map<String, MerkleEntityId> txHashToEntityId = new HashMap<>(); // HashMap<hash(txBytes), MerkleEntityId>
+
+	ScheduleID pendingId = NO_PENDING_ID;
+	String pendingTxHash = null;
+	MerkleSchedule pendingCreation;
+
+	public HederaScheduleStore(
+			EntityIdSource ids,
+			Supplier<FCMap<MerkleEntityId, MerkleSchedule>> schedules
+	) {
+		super(ids);
+		this.schedules = schedules;
+		buildTxBodyMap(this.schedules); // TODO: rebuild HashMap<hash(txBytes), {MerkleEntityId, List<AccountID>}>
+	}
+
+	@Override
+	public MerkleSchedule get(ScheduleID id) {
+		throwIfMissing(id);
+
+		return pendingId.equals(id) ? pendingCreation : schedules.get().get(fromScheduleId(id));
+	}
+
+	private void throwIfMissing(ScheduleID id) {
+		if (!exists(id)) {
+			throw new IllegalArgumentException(String.format("No such schedule '%s'!", readableId(id)));
+		}
+	}
+
+	@Override
+	public boolean exists(ScheduleID id) {
+		return pendingId.equals(id) || schedules.get().containsKey(fromScheduleId(id));
+	}
+
+	@Override
+	public void apply(ScheduleID id, Consumer<MerkleSchedule> change) {
+		throwIfMissing(id);
+		var key = fromScheduleId(id);
+		var schedule = schedules.get().getForModify(key);
+		Exception thrown = null;
+		try {
+			change.accept(schedule);
+		} catch (Exception e) {
+			thrown = e;
+		}
+		schedules.get().replace(key, schedule);
+		txHashToEntityId.remove(hex(schedule.transactionBody()));
+		if (thrown != null) {
+			throw new IllegalArgumentException("Schedule change failed unexpectedly!", thrown);
+		}
+	}
+
+	@Override
+	public CreationResult<ScheduleID> createProvisionally(byte[] bodyBytes, AccountID payer, AccountID schedulingAccount, RichInstant schedulingTXValidStart, Optional<JKey> adminKey) {
+		var validity = accountCheck(payer, INVALID_SCHEDULE_PAYER_ID);
+		if (validity != OK) {
+			return failure(validity);
+		}
+		validity = accountCheck(schedulingAccount, INVALID_SCHEDULE_ACCOUNT_ID);
+		if (validity != OK) {
+			return failure(validity);
+		}
+
+		pendingId = ids.newScheduleId(schedulingAccount);
+		pendingTxHash = hex(bodyBytes);
+		pendingCreation = new MerkleSchedule(
+				bodyBytes,
+				EntityId.ofNullableAccountId(schedulingAccount),
+				schedulingTXValidStart
+		);
+		adminKey.ifPresent(pendingCreation::setAdminKey);
+		pendingCreation.setPayer(EntityId.ofNullableAccountId(payer));
+
+		return success(pendingId);
+	}
+
+	@Override
+	public ResponseCodeEnum addSigners(ScheduleID sID, Set<JKey> signers) {
+		var id = resolve(sID);
+		if (id == MISSING_SCHEDULE) {
+			return INVALID_SCHEDULE_ID;
+		}
+
+		var schedule = get(id);
+		if (schedule.isDeleted()) {
+			return SCHEDULE_WAS_DELETED;
+		}
+
+		for (JKey signer: signers) {
```

Comment:
instead of iterating, maybe we can do `schedule.addSigners(signers)` and under the hood we could do `this.signers.addAll`